### PR TITLE
feat: page stats, page-range, percentage threshold, JSON/JUnit reports, and CLI

### PR DESCRIPTION
**Pull request by dichovsky**
## Summary

Adds five additive, backward-compatible features to `pdf-visual-compare`, ported from ideas in [`jamesmontemagno/pdf-diff`](https://github.com/jamesmontemagno/pdf-diff) (see `docs/pdf-diff-feature-audit.md` for the full audit that selected them):

| Feature | What it adds |
|---|---|
| **Change-% stats** | Per-page `mismatchPercent` + a document-level `summary` (`ComparePdfSummary`: counts by status, `maxMismatchPercent`, `totalMismatchCount`) on `ComparePdfDetailedResult`. |
| **Page-range selection** | `ComparePdfOptions.pages` — a `"1-3,5,7"` spec or a `number[]`. Filters the plan after discovery (only selected pages render). A selection matching no rendered page **throws** instead of passing vacuously. |
| **Percentage threshold** | `compareThresholdPercent` + per-page `matchingThresholdPercent` (0–100). **OR semantics**: a page passes when within EITHER the pixel-count OR the percent threshold, so a percentage *relaxes* (never tightens) the existing pixel default. |
| **Report serializers** | `toJsonReport(result)` and `toJUnitReport(result)` — exported pure functions producing pretty JSON or a deterministic JUnit XML report (one `<testcase>` per page). |
| **CLI** | `npx pdf-visual-compare <actual> <expected> [flags]` — hand-rolled (no new prod deps), bin `out/cli.js`. Flags: `--threshold`, `--threshold-percent`, `--pages`, `--format text\|json\|junit`, `--out`, `--fail-on-diff`, `--`, `-h/--help`, `-v/--version`. Exit codes `0`/`1`/`2`. |

## Public API additions

- `ComparePdfOptions`: `compareThresholdPercent`, `pages`
- `PageExclusion`: `matchingThresholdPercent`
- `ComparePdfPageResult`: `mismatchPercent`, `thresholdPercent`
- `ComparePdfDetailedResult`: `compareThresholdPercent`, `summary`
- New exports: `ComparePdfSummary` type, `toJsonReport`, `toJUnitReport`
- New `package.json` `bin` → `out/cli.js`

## Backward compatibility

`comparePdf` / `comparePdfDetailed` behavior is unchanged when the new options are unset. The additions to *result* types are non-breaking for consumers (they read, not construct, these types). No new production dependencies.

## Test plan

- 11 new test files (`__tests__/20`–`30`): pure-helper unit tests, option validation, percentage/page-range behavior, serializers (incl. XML escaping + control-char stripping), and the CLI (parser, orchestrator with DI, version reader). Updated the `13` result-shape contract test and the published-artifact / consumer-type surface checks.
- `npm test` (full published pipeline: clean → lint → codemap:check → license → build → type/artifact checks → vitest) passes at **100% coverage** (538 statements, 398 branches, 99 functions, 529 lines).
- CLI smoke-tested as a real binary: `--version`, `--help`, equal→exit 0, differ+`--fail-on-diff`→exit 1, bad usage→exit 2, `--` separator, JUnit/JSON output, flag-shaped-value rejection.

## Reviews

Each feature was reviewed by a dedicated reviewer pass plus a final cross-cutting integration review; all CRITICAL/HIGH/MEDIUM findings were addressed (e.g. OR-semantics doc clarity, page-range resource/edge-case guards, CLI `--` separator + late-rejection handling, JUnit null-safety).

## Notes / follow-ups

- Entries are under CHANGELOG **`[Unreleased]`**; the version is intentionally **not** bumped (a minor `4.1.0` release decision is left to the maintainer).
- Possible follow-up: an integration test exercising `pages` + `compareThresholdPercent` together (each path is covered separately).

### CODE REVIEW

File: src/internal/mismatchStats.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * `mismatchPercent` and any percentage threshold decision.
+ */
+export function computeMismatchPercent(
+    mismatchCount: number,
+    actualWidth: number,
+    actualHeight: number,
+    expectedWidth: number,
+    expectedHeight: number,
+): number {
+    const totalPixels = Math.max(actualWidth * actualHeight, expectedWidth * expectedHeight);
```
Comment (dichovsky) [inline src/internal/mismatchStats.ts]: Percentage thresholds can fail even at 100% for pages with different aspect ratios. `comparePng` normalizes inputs to a `max(width) x max(height)` canvas before returning `mismatchCount`, but this divides by only `max(actualArea, expectedArea)`. For example, a 100x1 page compared to a 1x100 page produces 199 mismatches, and this reports `199%`, so `compareThresholdPercent: 100` still marks the page mismatched. Please base the denominator on the normalized comparison canvas (`Math.max(actualWidth, expectedWidth) * Math.max(actualHeight, expectedHeight)`) or otherwise clamp the reported/thresholded percent to 100.

Comment (dichovsky) [reply src/internal/mismatchStats.ts]: Fixed in a81363e. Confirmed in `png-visual-compare`'s `normalizeImages.js` (`width = Math.max(width1, width2)`, `height = Math.max(height1, height2)`) — both pages are extended onto that canvas before `mismatchCount` is counted. The denominator now uses `max(actualWidth, expectedWidth) * max(actualHeight, expectedHeight)` instead of `max(area)`, so the 100×1 vs 1×100 case reports 1.99% (was 199%) and a fully-differing canvas reports exactly 100%. Same-size pages (the common case) are unaffected. Added a differing-aspect-ratio regression test in `__tests__/20.internal.mismatch-stats.test.ts` and updated the JSDoc, README, and agent docs.

File: src/internal/comparePlannedPage.ts [hunk 2]
```diff
@@ -82,14 +87,28 @@ export function comparePlannedPage(
...
+        planEntry.actualPage.width,
+        planEntry.actualPage.height,
+        planEntry.expectedPage.width,
+        planEntry.expectedPage.height,
+    );
+    // A page passes when it is within EITHER the pixel-count threshold OR the configured
+    // percentage threshold. The percentage tolerance therefore relaxes, never tightens, the
+    // pixel-count default; when no percentage threshold is configured only the pixel check applies.
+    const withinPixels = mismatchCount <= threshold;
+    const withinPercent = thresholdPercent !== null && mismatchPercent <= thresholdPercent;
```
Comment (dichovsky) [inline src/internal/comparePlannedPage.ts]: The threshold decision uses the 4-decimal value returned by `computeMismatchPercent`, so rounding can turn a real difference into an exact `0%` match. For example, 1 mismatched pixel on a 5000x5000 canvas computes to 0.000004%, is reported as `0`, and therefore passes `compareThresholdPercent: 0` even when `compareThreshold` is also 0. This creates a false negative in the core comparison result. Please compare the threshold against the unrounded ratio and round only the public/reporting value (with a regression test for a nonzero mismatch that rounds to zero).

File: src/serialize/xml.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// XML 1.0 forbids these control characters entirely — they cannot even be escaped as character
+// references, so they are stripped to keep the output well-formed.
+const ILLEGAL_XML_CHARS = /[\x00-\x08\x0B\x0C\x0E-\x1F]/g;
```
Comment (dichovsky) [inline src/serialize/xml.ts]: This filter does not cover all characters forbidden by XML 1.0, despite the serializer promising well-formed JUnit output. In particular, `U+FFFE`/`U+FFFF` and unpaired UTF-16 surrogates survive; a caller-controlled page name containing `U+FFFF` makes `xmllint` reject the generated report with `Char 0xFFFF out of allowed range`. Please sanitize according to the full XML 1.0 `Char` production (or use an XML serializer that does), and add regression coverage for non-control invalid code points.

File: src/internal/comparePlannedPage.ts (hunk 2 above)
Comment (dichovsky) [reply src/internal/comparePlannedPage.ts]: Fixed in 7979916. Split `computeMismatchPercent` into `computeUnroundedMismatchPercent` (used for the pass/fail decision) and the existing rounded version (used only for the reported `mismatchPercent` value). `withinPercent` now compares against the unrounded value, so the 1-mismatched-pixel-on-a-5000x5000-canvas case (0.000004%) correctly fails `compareThresholdPercent: 0` instead of silently passing. Added a regression test reproducing your exact scenario in `__tests__/23.compare.percentage-threshold.test.ts`, plus unit tests for the new helper in `__tests__/20.internal.mismatch-stats.test.ts`. Updated the doc line in CLAUDE.md/AGENTS.md accordingly.

File: src/serialize/xml.ts (hunk 3 above)
Comment (dichovsky) [reply src/serialize/xml.ts]: Fixed in 830ab31. `ILLEGAL_XML_CHARS` now strips every code point XML 1.0's `Char` production forbids: C0 controls (unchanged), `U+FFFE`/`U+FFFF`, and unpaired UTF-16 surrogates (via a negative lookahead/lookbehind on the high/low surrogate ranges) — while leaving valid surrogate pairs (e.g. emoji) intact. Verified end-to-end against `xmllint` with a `U+FFFF` page name: reproduces the reported "Char 0xFFFF out of allowed range" failure on the old code, clean exit on the fix. Added regression tests in `__tests__/27.serialize.junit-report.test.ts` covering `escapeXmlText`/`escapeXmlAttribute` directly, a lone-surrogate-adjacent-to-a-valid-pair edge case, escaping-order composition, and an integration test through `toJUnitReport`.
